turbonomic/kubeturbo
# Support running CPU frequency job on Power linux

## Intention

This PR supports running CPU frequency job on Power linux. The following changes/enhancements are made:

* `kubeturbo` will discover the OS and Arch of each node based on the [well known](https://kubernetes.io/docs/reference/labels-annotations-taints/) labels on that node.
* `kubeturbo` will log the OS and Arch info of each node during each full discovery.
* If `kubeturbo` needs to get CPU frequency by job, it will only run the job on a whitelist OS and Arch combinations.
* The `DefaultCpufreqJobExcludeNodeLabels` is removed.
* The `--cpufreq-job-exclude-node-labels` option has been enhanced to support multiple labels with the same key but different values.
* Implement interface, abstract type, and concrete types to handle job command and job output parsing of different os/arch.

## Implementation
* Implement interface `iNodeCpuFrequencyGetter`, abstract type `NodeCpuFrequencyGetter`, and concrete types `LinuxPpc64leNodeCpuFrequencyGetter` and `LinuxAmd64NodeCpuFrequencyGetter`. It is easy to extend and support other os/arch implementations in the future when needed.
* Implement the `GetNodeOSArch` method to parse os/arch from node label.
* Implement the helper function `KubeletClient.GetCpuFrequencyFromJob` which instantiates proper cpu frequency getter of a node based on os/arch of that node.
* Implement the utility function `LabelMapFromNodeSelectorString` to support exclusion labels with the same key but different values. For example, if user specifies 
`--cpufreq-job-exclude-node-labels="kubernetes.io/arch=s390x,kubernetes.io/arch=arm64"`, then both `s390x` and `arm64` node can be excluded (this is just an example, in practice this is not necessary as we only support two os/arch by default: `linux/amd64` and `linux/ppc64le`).

## Test
* Implement proper unit tests
* Build and deploy `kubeturbo` in an IBM Power Linux OCP cluster.
*  Check the log: 
```
I0811 18:10:14.523226       1 kubelet_client.go:306] Node rdr-instana-2-mon01-master-1 is running with linux/ppc64le.
...
I0811 18:10:14.523444       1 kubelet_client.go:306] Node rdr-instana-2-mon01-worker-0 is running with linux/ppc64le.
I0811 18:10:14.523517       1 kubelet_client.go:306] Node rdr-instana-2-mon01-master-0 is running with linux/ppc64le.
I0811 18:10:14.523576       1 kubelet_client.go:306] Node rdr-instana-2-mon01-master-2 is running with linux/ppc64le.
W0811 18:10:14.589974       1 kubelet_client.go:325] Failed to get CPU frequency for node rdr-instana-2-mon01-master-1 from kubelet: "https://192.168.26.12:10250/spec" was not found. Will try job based getter.
W0811 18:10:14.590079       1 kubelet_client.go:325] Failed to get CPU frequency for node rdr-instana-2-mon01-master-2 from kubelet: "https://192.168.26.108:10250/spec" was not found. Will try job based getter.
W0811 18:10:14.590479       1 kubelet_client.go:325] Failed to get CPU frequency for node rdr-instana-2-mon01-master-0 from kubelet: "https://192.168.26.127:10250/spec" was not found. Will try job based getter.
W0811 18:10:14.594946       1 kubelet_client.go:325] Failed to get CPU frequency for node rdr-instana-2-mon01-worker-0 from kubelet: "https://192.168.26.52:10250/spec" was not found. Will try job based getter.
I0811 18:10:20.674302       1 kubelet_client.go:388] CPU frequency of node rdr-instana-2-mon01-master-0: 2500 MHz
...
I0811 18:10:21.576702       1 kubelet_client.go:306] Node rdr-instana-2-mon01-worker-1 is running with linux/ppc64le.
W0811 18:10:21.579763       1 kubelet_client.go:325] Failed to get CPU frequency for node rdr-instana-2-mon01-worker-1 from kubelet: "https://192.168.26.207:10250/spec" was not found. Will try job based getter.
I0811 18:10:21.679130       1 kubelet_client.go:388] CPU frequency of node rdr-instana-2-mon01-master-1: 2500 MHz
I0811 18:10:21.685770       1 kubelet_client.go:388] CPU frequency of node rdr-instana-2-mon01-master-2: 2500 MHz
...
I0811 18:10:24.702581       1 kubelet_client.go:388] CPU frequency of node rdr-instana-2-mon01-worker-0: 2500 MHz
...
I0811 18:10:27.787162       1 kubelet_client.go:388] CPU frequency of node rdr-instana-2-mon01-worker-1: 2500 MHz
```
* **Before** the change, default `2000 MHz` is used 

  ![image](https://user-images.githubusercontent.com/10012486/129093883-2b9e9d3a-b0b0-4e4f-81bb-fa43f5880fa8.png)

* **After** the change, real `2500 MHz` is discovered

  ![image](https://user-images.githubusercontent.com/10012486/129094341-9418fb9d-a335-46f9-be8f-cc222d951a58.png)
